Lib/asyncio/timeouts.py

```diff
@@ -0,0 +1,182 @@
+import contextlib
+import enum
+
+from types import TracebackType
+from typing import final, Any, Iterator, Optional, Type
+
+from . import events
+from . import exceptions
+from . import tasks
+
+
+__all__ = (
+    "cancel_scope",
+    "cancel_scope_at",
+    "timeout",
+    "timeout_at",
+    "CancelScope"
+)
+
+
+class _State(enum.Enum):
+    CREATED = "created"
+    ENTERED = "active"
+    CANCELLING = "cancelling"
+    CANCELLED = "cancelled"
+    EXITED = "exited"
+
+
+@final
+class CancelScope:
+    __slots__ = ("_deadline", "_loop", "_state", "_task", "_timeout_handler")
+
+    def __init__(
+        self, deadline: Optional[float], loop: events.AbstractEventLoop
+    ) -> None:
+        self._loop = loop
+        self._state = _State.CREATED
+
+        self._timeout_handler: Optional[events.TimerHandle] = None
+        self._task: Optional[tasks.Task[Any]] = None
+        self._deadline = deadline
+
+    @property
+    def deadline(self) -> Optional[float]:
+        return self._deadline
+
+    @deadline.setter
+    def deadline(self, value: Optional[float]) -> None:
+        self._deadline = value
+        self._reschedule()
```

Comment:
I'm not keen on using properties for this kind of API.

Reply:
Is `scope.reschedule(deadline)` better?

Reply:
Use `reschedule()` method